discretisedfield/util/util.py

```diff
@@ -17,71 +18,89 @@ def array2tuple(array):
     return array.item() if array.size == 1 else tuple(array.tolist())
 
 
-def as_array(mesh, dim, val):
-    array = np.empty((*mesh.n, dim))
-    if isinstance(val, numbers.Real) and (dim == 1 or val == 0):
-        # The array for a scalar field with numbers.Real value or any
-        # field with zero value.
-        array.fill(val)
-    elif isinstance(val, complex) and (dim == 1 or val == 0):
-        array = np.empty((*mesh.n, dim), dtype='complex128')
-        array.fill(val)
-    elif isinstance(val, np.ndarray) and val.shape == array.shape:
-        array = val
-    elif isinstance(val, (tuple, list, np.ndarray)):
-        if len(val) != dim:
-            msg = (f'Wrong dimension {len(val)} provided for value;'
-                   f' expected dimension is {dim}')
-            raise ValueError(msg)
-        if (isinstance(val, np.ndarray) and val.dtype == 'complex128'
-                or np.iscomplex(val).any()):
-            array = np.empty((*mesh.n, dim), dtype='complex128')
+def as_array(mesh, dim, val, dtype=np.float64):
+    array = np.empty((*mesh.n, dim), dtype=dtype)
```

Comment:
`np.full` (https://numpy.org/doc/stable/reference/generated/numpy.full.html) creates and fills an array at the same time.

Reply:
Currently array creation and filling are separated (on purpose):
- Creation (i.e. the shape) only depends on the mesh and datatype.
- Filling has multiple options but is always operating on the same array (shape and dtype).
- Different input for `val` (scalar vs array-like) needs different types of error handling - and I'd like to avoid `if`-`else` in favour of registered functions.
- `np.full` will work for certain `val` (e.g. scalar, array-like) but not for `callable` and `dict`
- To avoid code duplication in creating the array I've therefore split it up. Furthermore, we can simply add new implementations for filling and know that we get an array of the correct shape.

If we want to use `np.full` we should at least remove `_fill_array` and directly overload `as_array`.

Reply:
I think it would be clearer if we directly overloaded `as_array`.